### ISSUE TITLE
Implement new actions for sharing settings

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#4794]
## Feature Description

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

The `core/modules` store should be extended with the following actions:

* `setSharingManagement(moduleSlug: string, management: enum)`  
Sets the value for a module's sharing management where `management` can be one of `all_admins` or `owner`
* `setSharedRoles(moduleSlug: string, roles: string[])`  
Sets the list of roles (`role.id`[]) the module is shared with
* `saveSharingSettings()`  
Saves all sharing settings via `POST:core/modules/data/sharing-settings` using `createFetchStore`
	- See https://github.com/google/site-kit-wp/issues/4481 for the response shape
	- The current and saved settings should be received/refreshed in the state from the response
	- Any `newOwnerIDs` included in the response should update the modules in state
* The state for settings should work similar to `createSettingsStore` which stores state for saved settings separately from the current working/draft values that selectors and actions touch
	- Accordingly, an action for `receiveGetSharingSettings` should also be created which should update both
* Test coverage should be added for all new actions

## Implementation Brief
* Create `assets/js/googlesitekit/modules/datastore/sharing-settings.js` which will contain our actions and reducers.
* The `baseInitialState` should contain the `sharingSettings` and `savedSharingSettings` property, both initialized to an empty object.
* Create `fetchSaveSharingSettingsStore` using the `createFetchStore` function with the following details:
	* `baseName` should be set to `saveSharingSettings`
	* `controlCallback` should:
		* require one parameter, an object having the same shape as `state.savedSharingSettings`.
		* make a POST request to `core/modules/data/sharing-settings` using `API.set`, and passing the `state.savedSharingSettings` as the body.
		* the `argsToParams` and `validateParams` should be set as well. Refer to how it's done in other data partials.
		* For the `reducerCallback`, 
			* `state.sharingSettings` should be updated to the response from the API, i.e both `state.sharingSettings` and `state.saveSharingSettings` should be equal.
			* check if `newOwnerIDs` from the API response is not empty. Refer to https://github.com/google/site-kit-wp/issues/4481 for the structure. In this case, loop through the array and update each module owner ID using the `setOwnerID` action.
* Similar to other data store partials in the folder, add the following `actions` within the `baseAction` object:
	* `setSharingManagement`
		* The constant `SET_SHARING_MANAGEMENT` having the same value should be added at the beginning of the file.
		* It accepts the `moduleSlug` and `management` as mandatory parameters. 
		* The corresponding reducer should be added to the reducer function where the `management` property of the given `moduleSlug` within the state `savedSharingSettings` property.
	* `setSharedRoles`
		* The constant `SET_SHARED_ROLES` having the same value should be added at the beginning of the file.
		* It accepts the `moduleSlug` and `roles` as mandatory parameters.
		* The corresponding reducer should be added to the `baseReducers` object where the `sharedRoles` property of the given `moduleSlug` is set within the state `savedSharingSettings` property.
	* `saveSharingSettings`
		* uses the `createValidatedAction` function which checks whether `state.savedSharingSettings` is not empty.
		* The function should call the `fetchSaveSharingSettingsStore.actions.fetchSaveSharingSettings` function, passing the `state.savedSharingSettings`.
	* `receiveGetSharingSettings`
		* It should accept a parameter, similar to `settings` and it should set `state.sharingSettings` and `state.savedSharingSettings` with the provided parameter.
* For the newly added actions,
	* use `invariant` to check for the parameters similar to how it’s done in other data store partials.
	* Refer to the AC for the parameter types.
* Update `assets/js/googlesitekit/modules/datastore/index.js` to include the new data store partial.

### Test Coverage
* Test coverage should be added for all new actions, selectors .

## QA Brief

**QA**

**QA:Eng**

- Ensure the `dashboardSharing` feature flag is enabled.
- In the console, run the `receiveGetSharingSettings` action like so:
```js
googlesitekit.data.dispatch('core/modules').receiveGetSharingSettings({
	'search-console': {
		sharedRoles: [ 'editor', 'subscriber' ],
		management: 'all_admins',
	},
	analytics: {
		sharedRoles: [ 'editor' ],
		management: 'owner',
	},
	'pagespeed-insights': {
		sharedRoles: [ 'editor' ],
		management: 'all_admins',
	},
})
```
- Note that the `receiveGetSharingSettings` action has to be dispatched before `saveSharingSettings` since it sets the `sharingSettings` in the state. Otherwise, it will throw an error.
	- This is only necessary until #4795 is merged
- In the console, test invoking the added actions as defined in the ACs  
    ```js
	await googlesitekit.data.dispatch('core/modules')...
	```

- Test `saveSharingSettings` makes an API call to the `sharing-settings` endpoint successfully

## Changelog entry

* Implement new actions for sharing settings.

**Comment by hussain-t** [comment]
@asvinb @aaemnnosttv The `getSharingSettings()` selector is being implemented in #4795.
>Add the getSharingSettings selector which returns state.settings

**Comment by tofumatt** [comment]
IB ✅

**Comment by aaemnnosttv** [comment]
@kuasha420 @hussain-t This issue should not depend on #4481. The two are interconnected but should be able to be implemented separately. I'm going to remove the dependency for now but let me know if you feel it is important to keep for some reason.

**Comment by hussain-t** [comment]
@aaemnnosttv, there are a couple of reasons I consider #4481 is blocking this ticket though I have done the base work with unit tests and created a draft PR. 
1. I wanted to know the exact shape of the sharing settings 
2. I want to test the `saveSharingSettings` action locally and for QA testing.

**Comment by hussain-t** [comment]
@aaemnnosttv, we need to move the `getSharingSettings` selector and resolver from #4795 to this ticket since we need to get the `sharingSettings` from the `getSharingSettings` selector to use it in the `saveSharingSettings` action to pass it to the `fetchSaveSharingSettings` action like this:

<img width="726" alt="Screenshot 2022-05-16 at 3 06 40 PM" src="https://user-images.githubusercontent.com/24408261/168563995-8d4649cf-f45c-492e-bd50-160a3a1895ce.png">

**Comment by aaemnnosttv** [comment]
> we need to move the `getSharingSettings` selector and resolver from #4795 to this ticket since we need to get the `sharingSettings` from the `getSharingSettings` selector to use it in the `saveSharingSettings` action to pass it to the `fetchSaveSharingSettings` action like this:

@hussain-t I would prefer to keep these in their respective issues as defined and use a temporary workaround here instead.

E.g. we can manually "select" the state for the action above instead, and refactor it to use the selector in that issue as a simple one line change.

```js
// TODO: Refactor to use `getSharingSettings` selector.
const { sharingSettings } = registry.stores[ CORE_MODULES ].store.getState()
```

That should unblock this issue and then we can address the above TODO in #4795. If that works for you, then let's move forward with that approach and we can simply update the IB for 4795 with a reminder to address this TODO (I don't think any changes to ACs are necessary here).

**Comment by hussain-t** [comment]
Thanks, @aaemnnosttv, SGTM 👍

**Comment by hussain-t** [comment]
Thanks, @aaemnnosttv, as per our discussion I have updated the tests to dispatch the `receiveGetSharingSettings` action then we should be able to get the `sharingSettings` from the `*.store.getState()`. I have also updated the QAB respectively.

**Comment by aaemnnosttv** [comment]
## ⚠️ QA:Eng

Somewhat of a secondary CR here but a few minor things should be fixed.

@hussain-t please see my comments here: https://github.com/google/site-kit-wp/pull/5127#pullrequestreview-983644221

**Comment by hussain-t** [comment]
Thanks, @aaemnnosttv. I have addressed your comments in a [follow-up PR](https://github.com/google/site-kit-wp/pull/5262) and assigned this ticket for review.

**Comment by jimmymadon** [comment]
## QA: Eng ✅ 

- Verified the new actions within the ACs using the recently edited QA brief.

Screenshots:

<details>
 <summary>setSharingManagement</summary>

<img width="1029" alt="Screenshot 2022-06-05 at 18 37 24" src="https://user-images.githubusercontent.com/1721319/172063134-bc26c9b4-521b-4656-b95c-d83407fb60b2.png">

<img width="1033" alt="Screenshot 2022-06-05 at 18 36 25" src="https://user-images.githubusercontent.com/1721319/172063169-6987cd83-f169-4535-a79f-c1fd3ae437cf.png">

<img width="1030" alt="Screenshot 2022-06-05 at 18 36 04" src="https://user-images.githubusercontent.com/1721319/172063165-8886935c-4ad8-4389-964e-13b27f38d895.png">

</details>

<details>
 <summary>setSharedRoles</summary>
<img width="1013" alt="Screenshot 2022-06-05 at 19 18 06" src="https://user-images.githubusercontent.com/1721319/172063234-25b0adab-3001-4c6a-856a-97fd5dc36ad6.png">

<img width="1031" alt="Screenshot 2022-06-05 at 19 18 54" src="https://user-images.githubusercontent.com/1721319/172063247-a5e47795-c065-461e-ab8e-e88ce49aaf92.png">
</details>

<details>
 <summary>saveSharingSettings</summary>

<img width="586" alt="Screenshot 2022-06-05 at 19 23 24" src="https://user-images.githubusercontent.com/1721319/172063272-8a787aed-5ce4-41c7-affa-82d193d4b166.png">

<img width="1035" alt="Screenshot 2022-06-05 at 19 22 28" src="https://user-images.githubusercontent.com/1721319/172063284-a4b6a88d-c18f-47a3-89ad-753ca63efffa.png">

</details>
